### ISSUE TITLE
Ability to select logging to stderr

### ISSUE THREAD

**Issue by Veetaha** [borntyping/rust-simple_logger#21]
It would be awesome if it was possible to switch from the hardcoded `println!` to some user-specified output destination.
For my particular use-case, I'd like for the logs to go to `stderr`, but, I think, this should not be difficult to generalize this further to any `impl std::io::Write` type?

**Comment by piegamesde** [comment]
I just gave this a shot, and it's actually not that easy. The type system requirements are:

- Use a boxed trait object for the writer, as we probably don't want to expose this as a generic parameter (also we probably can't with the current builder pattern implementation)
- `impl Write` is not enough. Due to multithreading, we're at `Write + Send + Sync` which I'm not sure how many writers except `Stdout` and `Stderr` implement.
- Also we can't modify the `SimpleLogger` struct after initialization, but `Write` requires a mutable borrow on the writer. Which leaves us with a `RefCell`.
- Now I've reached a `RefCell<Box<dyn Write + Send + Sync>>` and of course this isn't thread safe anymore so I'd need to wrap this in a `Mutex` or some other concurrency primitive. Which makes painfully clear that this would synchronize all threads on each log call, so I doubt this is a good approach at the problem at all.

-----

Let's look at other approaches.

A simple switch for `println!` vs `eprintln!` is pretty easy to implement, that's what [EnvLogger does](https://docs.rs/env_logger/0.7.1/env_logger/enum.Target.html).

`simplelog` uses [a Mutex and a generic parameter](https://docs.rs/simplelog/0.8.0/src/simplelog/loggers/writelog.rs.html#17-21).

**Comment by Veetaha** [comment]
@piegamesde thank you for the involvement. This does require some internal mutability, and I assure you neither `RefCell` nor `Mutex` are a bad thing, it's important to evaluate the options, and you've already shown some of them.

I recommend taking a look at `println!` internals. At first glance it uses some [`thread_local!` `RefCell<Option<Box<dyn Write + Send>>>`](https://github.com/rust-lang/rust/blob/5099914a16a215794ad243df0cc7a05d91d168e0/library/std/src/io/stdio.rs#L17-L22), but it is initialized with `None` and I guess is never overwritten, because it is exposed only via an [unstable function `set_print`](https://github.com/rust-lang/rust/blob/5099914a16a215794ad243df0cc7a05d91d168e0/library/std/src/io/stdio.rs#L841-L862). So it ignores this thread-lockal value in [`print_to`](https://github.com/rust-lang/rust/blob/5099914a16a215794ad243df0cc7a05d91d168e0/library/std/src/io/stdio.rs#L909-L911) and does use a global stdout lock with a [`ReentrantMutex`](https://github.com/rust-lang/rust/blob/5099914a16a215794ad243df0cc7a05d91d168e0/library/std/src/io/stdio.rs#L537-L540) by calling [`std::io::stdout()`](https://github.com/rust-lang/rust/blob/5099914a16a215794ad243df0cc7a05d91d168e0/library/std/src/io/stdio.rs#L893) function.
Inside of its `io::Write` impl `Stdout` locks this reentrant mutex on [each write](https://github.com/rust-lang/rust/blob/5099914a16a215794ad243df0cc7a05d91d168e0/library/std/src/io/stdio.rs#L766).

So, in conclusion, I'll say that `Mutex/ReentrantMutex` is not a bad thing, I don't see other means of sharing a global mutable io stream. We might try to copy the setup used in `stdout`, maybe use an enum like this to support already `Sync` types:
```rust
struct ConcurrentWriter {
     Unsync(ReentrantMutex<Box<dyn Write + Send>>),
     Sync(Box<dyn Write + Send + Sync>)
}
```
though I see this is going to be quite complicated for a `simple_logger`.
Taking this into account, I think, it does make sense just to add a fixed switch between `stdout/stderr` via a boolean setter on the builder, or via a conditional compilation (cargo feature like `features = ["stderr"]`), though I'd just go with a boolean setter.

**Comment by piegamesde** [comment]
Thank you for looking that up! I didn't know stdout did some synchronizing between threads as well ... I still think it is possible to do "unlocked" multithreaded logging, but it would at a lot of complexity and probably little benefit to the code.

In order to keep things "simple", I propose a solution similar to `fern`s, but with less variants:

```
enum Output {
    Stdout,
    Stderr,
}
```

we keep the enum private in order to have the possibility to add file writers later on (if desired).

We could also store the `Stdout` and `Stderr` struct in the respective variants, but in the end it should be equivalent to using `println` and `eprintln` directly, so I don't have any preference on that one.

**Comment by borntyping** [comment]
A very basic version of this is availible in [1.13.0](https://github.com/borntyping/rust-simple_logger/releases/tag/v1.13.0) as a compile time feature. I may interate on this in the future but it likely won't be any more complex than a toggle between `println` and `eprintln`.

**Comment by cehteh** [comment]
albeit closed now, I want to note that it would be simple to add a 'channel' target for logging (features = "channel").  Then provide a API to get the reading end of the channel and the user has to implement the IO on its own. This solves the problems with the locking mentioned above.

Implementations are then choose between dedicated threads, async or whatever floats the boat. Targets like files, syslog and anything else become possible this way without compromise on the 'simplicity' of this crate.

I have not need for that yet, but if acceptable I may someday provide an implementation for that.
